### PR TITLE
Framework: Prevent JS script loads from blocking render

### PR DESCRIPTION
**Pull request by seear**
Action from p55Cj4-lD-p2

Add `defer` attr to script tags for app js files. This will allow initial html render to proceed before scripts have loaded, which will decrease time to render for logged-out server-rendered pages.

**This PR adds defer attribute in development and wpcalypso environments**

**To Test**
* `make run` or `NODE_ENV=wpcalypso make run` should have `defer` in the <script> tags in the page source
* other envs, for example 'horizon', should be as `master`
* Building with `code-splitting: false` should have the `onload` attr on the `build...js` rather than the chunk script

### CODE REVIEW

**Comment by seear** [comment]
To test an SSR'd page live (logged-out) https://calypso.live/themes?branch=add/framework-js-defer

**Comment by seear** [comment]
Waterfall charts from WebPageTest for _/themes_ on a couple of calypso.live branches. Calypso.live is slow, so disregard the absolute numbers. The key comparison is the vertical green *start render* line.

I used another PR with only a small css change as a baseline:
screen shot 2017-04-11 at 14 49 09
https://wpt.a8c.com/result/170411_81_133/1/details/

And on this branch:
screen shot 2017-04-11 at 14 48 49
https://wpt.a8c.com/result/170411_CX_131/1/details/

Whilst the DOM Content Loaded times are roughly the same, we can see that the *start render* line is much earlier on this branch, before some of the js files have even started downloading.

**Comment by seear** [comment]
**Questions**

*Should `defer` only be used for logged-out, SSRd pages?*
Since the base html is minimal for logged-in, it probably doesn't matter if the render goes ahead before the js loads.

*Can we rely on `onload='window.AppBoot()` on the chunk js script in all scenarios, instead of the inline script we had previously?*

**Comment by seear** [comment]
> Can we rely on onload='window.AppBoot() on the chunk js script in all scenarios, instead of the inline script we had previously?

Changed code in 25011ed so the `onload` is added to the last script.

File: server/pages/index.jade [hunk 1]
```diff
@@ -129,18 +129,28 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
...
-			script(src=urls[ jsFile ])
+			script(defer, src=urls[ 'manifest' ])
+			script(defer, src=urls[ 'vendor' ])
 			if chunk
-				script(src=urls[ chunk ])
+				script(defer, src=urls[ jsFile ])
+				script(defer, src=urls[ chunk ] onload='window.AppBoot()')
+			else
+				script(defer, src=urls[ jsFile ] onload='window.AppBoot()')
+		else if 'wpcalypso' === env
```
Comment (seear) [inline server/pages/index.jade]: Final, production, PR would remove this section.

Comment (gziolo) [inline server/pages/index.jade]: Good idea 👍

**Comment by ehg** [comment]
Cool, here's Lighthouse logged-out /themes before this PR — will comment with another when it's merged :) 

[lighthouse before defer.pdf](https://github.com/Automattic/wp-calypso/files/916279/lighthouse.before.defer.pdf)

**Comment by seear** [comment]
Thanks for reviews!

**Comment by seear** [comment]
WPT waterfalls for wpcalypso.wordpress.com/themes

Before merge:
screen shot 2017-04-12 at 12 45 15

After merge:
screen shot 2017-04-12 at 12 45 26

Looks like it made no difference. Looking again at the waterfall for for wordpress.com/themes, it looks to me like render start lines up more closely with CSS load finish and first screenshot load:
screen shot 2017-04-12 at 12 49 05

**Comment by ehg** [comment]
[lighthouse after defer.pdf](https://github.com/Automattic/wp-calypso/files/916933/lighthouse.after.defer.pdf)

> render start lines up more closely with CSS load finish

Yes, we're blocked by noticons, dashicons and our monolithic CSS bundle here.

**Comment by gziolo** [comment]
still 1 point earned 🏅

**Pull request merged**